backtrace-library/src/androidTest/java/backtraceio/library/BacktraceClientMetricsTest.java

```diff
@@ -0,0 +1,950 @@
+    }
+
+    @Test
+    public void shouldUploadEventsWhenMaxNumEventsReached() {
+        final int maximumNumberOfEvents = 3;
+        final Waiter waiter = new Waiter();
+
+        backtraceClient.metrics.enable(new BacktraceMetricsSettings(universeName, token, defaultBaseUrl, 0));
+        backtraceClient.metrics.setMaximumNumberOfEvents(maximumNumberOfEvents);
+        final MockRequestHandler mockRequestHandler = new MockRequestHandler();
+        backtraceClient.metrics.setUniqueEventsRequestHandler(mockRequestHandler);
+
+        backtraceClient.setUniqueEventsOnServerResponse(new EventsOnServerResponseEventListener() {
+            @Override
+            public void onEvent(EventsResult result) {
+                assertEquals(0, result.getEventsPayload().getDroppedEvents());
+                String eventsPayloadJson = BacktraceSerializeHelper.toJson(result.getEventsPayload());
+                assertNotEquals(0, eventsPayloadJson.length());
+                assertEquals(BacktraceResultStatus.Ok, result.status);
+                waiter.resume();
+            }
+        });
+
+        for (int i = 0; i < maximumNumberOfEvents; i++) {
+            backtraceClient.metrics.addUniqueEvent(uniqueAttributeName[i]);
+        }
+
+        try {
+            waiter.await(1000);
+        } catch (Exception e) {
+            fail(e.toString());
+        }
+
+        assertEquals(1, mockRequestHandler.numAttempts);
+        assertNotEquals(0, mockRequestHandler.lastEventPayloadJson.length());
+        assertEquals(maximumNumberOfEvents, backtraceClient.metrics.getUniqueEvents().size());
+    }
+
+    @Test
+    public void shouldNotUploadEventsBeforeMaxNumEventsReached() {
+        final int maximumNumberOfEvents = 3;
+        final Waiter waiter = new Waiter();
+
+        backtraceClient.metrics.enable(new BacktraceMetricsSettings(universeName, token, defaultBaseUrl, 0));
+        backtraceClient.metrics.setMaximumNumberOfEvents(maximumNumberOfEvents);
+        final MockRequestHandler mockRequestHandler = new MockRequestHandler();
+        backtraceClient.metrics.setUniqueEventsRequestHandler(mockRequestHandler);
+
+        backtraceClient.setUniqueEventsOnServerResponse(new EventsOnServerResponseEventListener() {
+            @Override
+            public void onEvent(EventsResult result) {
+                assertEquals(0, result.getEventsPayload().getDroppedEvents());
+                String eventsPayloadJson = BacktraceSerializeHelper.toJson(result.getEventsPayload());
+                assertNotEquals(0, eventsPayloadJson.length());
+                assertEquals(BacktraceResultStatus.Ok, result.status);
+                waiter.resume();
+            }
+        });
+
+        // Account for unique events startup event
+        for (int i = 0; i < maximumNumberOfEvents - 2; i++) {
+            backtraceClient.metrics.addUniqueEvent(uniqueAttributeName[0]);
+        }
+
+        try {
+            sleep(100);
```

Comment:
why we have this sleep(...)?

Reply:
@BartoszLitwiniuk we want to make sure we DON'T send something, but we have to wait because if it does send something (test fails) it will be delayed

Reply:
I think this is actually useless, I will remove this, thank you!

Reply:
Did we change it ? I though it will be better if we will directly invoke send API from there to avoid sleeps etc.

Reply:
We did change this, see the new `disableAutoSend` test. I don't know why it doesn't show as `Outdated` on GitHub but it is changed now.